etsin_finder/frontend/js/stores/view/common.files.js

```diff
@@ -0,0 +1,347 @@
+import axios from 'axios'
+import { observable, action, runInAction, computed } from 'mobx'
+
+import { FILE_API_URLS } from '../../utils/constants'
+import { Project, dirIdentifierKey, fileIdentifierKey } from './common.files.items'
+import { PromiseManager } from './common.files.utils'
+import { itemLoaderPublic } from './common.files.loaders'
+
+class Files {
+  // Base class for file hierarchies.
+
+  constructor() {
+    this.promiseManager = new PromiseManager()
+    Files.prototype.reset.call(this)
+  }
+
+  @observable datasetIdentifier = null
+
+  @observable root = null
+
+  @observable selectedProject = undefined
+
+  @observable projectLocked = false // prevent changing saved project
+
+  @observable loadingProjectInfo = null
+
+  @observable loadingProjectRoot = null
+
+  @observable loadingMetadata = null
+
+  cache = {} // used for storing data for items that haven't been fully loaded yet
+
+  @observable originalMetadata = {}
+
+  @action reset() {
+    this.datasetIdentifier = null
+    this.root = null
+    this.selectedProject = undefined
+    this.refreshModalDirectory = null
+    this.cache = {}
+    this.originalMetadata = {}
+    this.projectLocked = false
+
+    if (this.loadingProjectInfo && this.loadingProjectInfo.promise) {
+      this.loadingProjectInfo.promise.cancel()
+    }
+    this.loadingProjectInfo = null
+
+    if (this.loadingMetadata && this.loadingMetadata.promise) {
+      this.loadingMetadata.promise.cancel()
+    }
+    this.loadingMetadata = null
+
+    if (this.loadingProjectRoot && this.loadingProjectRoot.promise) {
+      this.loadingProjectRoot.promise.cancel()
+    }
+    this.loadingProjectRoot = null
+    this.promiseManager.reset()
+  }
+
+  @action loadProjectInfo = async () => {
+    const identifier = this.datasetIdentifier
+    const alreadyLoading = this.loadingProjectInfo && this.loadingProjectInfo.identifier === this.identifier
+    if (alreadyLoading) {
+      return
+    }
+
+    if (this.loadingProjectInfo && this.loadingProjectInfo.promise && !this.loadingProjectInfo.error) {
```

Comment:
Like alkreadyLoading above these checks could be named eg. returnedError

Reply:
done